src/dbnode/storage/series/buffer.go

```diff
@@ -181,287 +169,271 @@ func (b *dbBuffer) Write(
 	}
-	return result
-}
 
-func (b *dbBuffer) Snapshot(ctx context.Context, blockStart time.Time) (xio.SegmentReader, error) {
-	var (
-		res xio.SegmentReader
-		err error
-	)
+	// A call to flush can only be for warm writes.
+	blocks, err := buckets.ToBlocks(WarmWrite)
+	if err != nil {
+		return FlushOutcomeErr, err
+	}
 
-	b.forEachBucketAsc(func(bucket *dbBufferBucket) {
-		if err != nil {
-			// Something already went wrong and we want to return the error to the caller
-			// as soon as possible instead of continuing to do work.
-			return
+	// In the majority of cases, there will only be one block to persist here.
+	// Only when a flush fails
+	for _, block := range blocks {
+		if block == nil {
+			return FlushOutcomeBlockDoesNotExist, nil
 		}
 
-		if !bucket.canRead() {
-			return
+		stream, err := block.Stream(ctx)
+		if err != nil {
+			return FlushOutcomeErr, err
 		}
 
-		if !blockStart.Equal(bucket.start) {
-			return
+		segment, err := stream.Segment()
+		if err != nil {
+			return FlushOutcomeErr, err
 		}
 
-		// We need to merge all the bootstrapped blocks / encoders into a single stream for
-		// the sake of being able to persist it to disk as a single encoded stream.
-		_, err = bucket.merge()
+		checksum, err := block.Checksum()
 		if err != nil {
-			return
+			return FlushOutcomeErr, err
 		}
 
-		// This operation is safe because all of the underlying resources will respect the
-		// lifecycle of the context in one way or another. The "bootstrapped blocks" that
-		// we stream from will mark their internal context as dependent on that of the passed
-		// context, and the Encoder's that we stream from actually perform a data copy and
-		// don't share a reference.
-		streams := bucket.streams(ctx)
-		if len(streams) != 1 {
-			// Should never happen as the call to merge above should result in only a single
-			// stream being present.
-			err = errMoreThanOneStreamAfterMerge
-			return
+		err = persistFn(id, tags, segment, checksum)
+		if err != nil {
+			return FlushOutcomeErr, err
 		}
+	}
 
-		// Direct indexing is safe because canRead guarantees us at least one stream
-		res = streams[0]
-	})
+	if bucket, exists := buckets.WritableBucket(WarmWrite); exists {
+		bucket.SetVersion(version)
+	}
 
-	return res, err
+	return FlushOutcomeFlushedToDisk, nil
 }
 
 func (b *dbBuffer) ReadEncoded(ctx context.Context, start, end time.Time) [][]xio.BlockReader {
 	// TODO(r): pool these results arrays
 	var res [][]xio.BlockReader
-	b.forEachBucketAsc(func(bucket *dbBufferBucket) {
-		if !bucket.canRead() {
-			return
-		}
-		if !start.Before(bucket.start.Add(b.blockSize)) {
-			return
-		}
-		if !bucket.start.Before(end) {
+
+	b.forEachBucketAsc(func(bv BufferBucketVersions) {
+		if !bv.StartTime().Before(end) || !start.Before(bv.StartTime().Add(b.blockSize)) {
 			return
 		}
 
-		res = append(res, bucket.streams(ctx))
+		res = append(res, bv.Streams(ctx, WarmWrite))
+		res = append(res, bv.Streams(ctx, ColdWrite))
```

Comment:
Hm if there are no cold writes, wouldn't this put nil here?  Probably want to avoid for that case I assume?